cabal-install/src/Distribution/Client/Config.hs

```diff
@@ -587,40 +590,87 @@ initialSavedConfig = do
     }
   }
 
-defaultCabalDir :: IO FilePath
-defaultCabalDir = getAppUserDataDirectory "cabal"
-
-getCabalDir :: IO FilePath
-getCabalDir = do
+-- | If @CABAL\_DIR@ is set or @~/.cabal@ exists, return that
+-- directory.
+maybeGetCabalDir :: IO (Maybe FilePath)
```

Comment:
A better name for this function would be `maybeGetDotCabalDir`, because it’s specifically ~/.cabal that it tries to find, not any $CABAL_DIR or anything.

Reply:
cabal-install seems to use the term "cabal dir" for "the directory that contains all cabal-related stuff", i.e. the directory that serves the same purpose as the old `~/.cabal`.  Whether it actually has a leading dot in its name is immaterial.  I tried to preserve that nomenclature.

Reply:
Ok, now I saw I was wrong it checks all sorts of things, not just ~/.cabal. In that case, I'm confused as to why it's a Maybe. Could  you document when it can give you a nothing? And maybe some more prose as to what our tried to achieve. The current comment seems completely wrong?

Reply:
I added a clarification.

Reply:
Big improvement, thank you! I'm still not completely sure about the name of the function. How about: `maybeGetAllInOneCabalDir` or `maybeGetPreXdgCabalDir`?

Reply:
This is ultimately subjective, and doesn't really matter.  "cabal dir" is a fairly common term of art in the code base, so I'm sticking with that for now.